extensions/amp-ad/0.1/amp-ad.js

```diff
@@ -0,0 +1,403 @@
+    }
+    return cidForOrNull(this.getWin()).then(cidService => {
+      if (!cidService) {
+        return;
+      }
+      let consent = Promise.resolve();
+      if (consentId) {
+        consent = userNotificationManagerFor(this.getWin()).then(service => {
+          return service.get(consentId);
+        });
+        if (!scope && consentId) {
+          return consent;
+        }
+      }
+      return cidService.get(scope, consent);
+    });
+  }
+
+  /** @override  */
+  viewportCallback(inViewport) {
+    if (this.intersectionObserver_) {
+      this.intersectionObserver_.onViewportCallback(inViewport);
+    }
+    this.sendEmbedInfo_(inViewport);
+  }
+
+  /**
+   * @param {boolean} inViewport
+   * @private
+   */
+  sendEmbedInfo_(inViewport) {
+    if (this.iframe_) {
+      const targetOrigin =
+          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
+      postMessage(this.iframe_, 'embed-state', {
+        inViewport: inViewport,
+        pageHidden: !this.viewer_.isVisible(),
+      }, targetOrigin, /* opt_is3P */ true);
+    }
+  }
+
+  /** @override  */
+  overflowCallback(overflown, requestedHeight, requestedWidth) {
+    if (overflown) {
+      const targetOrigin =
+          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
+      postMessage(
+          this.iframe_,
+          'embed-size-denied',
+          {requestedHeight: requestedHeight, requestedWidth: requestedWidth},
+          targetOrigin,
+          /* opt_is3P */ true);
+    }
+  }
+
+  /**
+   * Updates the element's dimensions to accommodate the iframe's
+   *    requested dimensions.
+   * @param {number|undefined} newWidth
+   * @param {number|undefined} newHeight
+   * @private
+   */
+  updateSize_(newHeight, newWidth) {
+    this.attemptChangeSize(newHeight, newWidth, () => {
+      const targetOrigin =
+          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
+      postMessage(
+          this.iframe_,
+          'embed-size-changed',
+          {requestedHeight: newHeight, requestedWidth: newWidth},
+          targetOrigin,
+          /* opt_is3P */ true);
+    });
+  }
+
+  /**
+   * Activates the fallback if the ad reports that the ad slot cannot
+   * be filled.
+   * @private
+   */
+  noContentHandler_() {
+    // If a fallback does not exist attempt to collapse the ad.
+    if (!this.fallback_) {
+      this.attemptChangeHeight(0, () => {
+        this.element.style.display = 'none';
+      });
+    }
+    this.deferMutate(() => {
+      if (this.fallback_) {
+        // Hide placeholder when falling back.
+        if (this.placeholder_) {
+          this.togglePlaceholder(false);
+        }
+        this.toggleFallback(true);
+      }
+      // Remove the iframe only if it is not the master.
+      if (this.iframe_.name.indexOf('_master') == -1) {
+        this.element.removeChild(this.iframe_);
+        this.iframe_ = null;
+      }
+    });
+  }
+}
+AMP.registerElement('amp-ad', AmpAd);
```

Comment:
AMP.registerElement() calls registerExtendedElement() in src/runtime.js and registerExtendedElement() calls upgradeOrRegisterElement() which can be found in src/extended-element.js. 
It still works when called twice I think.